src/torchmetrics/functional/retrieval/average_precision.py

```diff
@@ -40,10 +47,14 @@ def retrieval_average_precision(preds: Tensor, target: Tensor) -> Tensor:
     """
     preds, target = _check_retrieval_functional_inputs(preds, target)
 
+    top_k = top_k or preds.shape[-1]
+    if not isinstance(top_k, int) and top_k <= 0:
```

Comment:
Why limit to `int` here? Theoretically something like `for k in torch.arange(10) ...` should also work with `torch.topk`.
What is the advantage to disallow it?

Reply:
what would be the return type here? I can see it for a specific top_k value, but how would it look like for multiple ks?

Reply:
I meant to ask about it as a design choice.
I'm guessing that the check was added since `torch.topk` wont accept `float`s as `k`:
```python
>>> x = torch.rand(10)
>>> x.topk(1.0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: topk(): argument 'k' (position 1) must be int, not float
```
`torch` throws a `TypeError` ([which is the correct exception](https://docs.python.org/3/library/exceptions.html#TypeError), `ValueError` actually is not fitting.) to tell the user what went wrong.

With our type checking, we are enforcing a stricter limitation than necessary. PyTorch allows other types that behave like `int`s, for example:
```python
>>> x = torch.rand(10)
>>> for k in np.arange(1, 10):
...     x.topk(k)
...
>>> for k in torch.arange(1, 10):
...     x.topk(k)
... 
>>> a = torch.as_tensor([True, False, True]).sum()
>>> x.topk(a)
...
```

The intent of the this [LBYL-style design](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/) was probably to help users of this library do the right thing but I think it is limiting instead.
I propose to remove the type check since it's simpler, more maintainable, doesn't get in the way, and torch will give a meaningful error message anyways.

As I said above, I think it's a design choice and I just wanted to share my view as a user of this library :)